File: src/plugins/lua/meta/qt.lua
---@meta Qt

--- The values in enums here do not matter, as they are defined by the C++ code.

local qt = {}

---@enum CompleterCompletionMode
qt.CompleterCompletionMode = {
    PopupCompletion = 0,
    InlineCompletion = 1,
    UnfilteredPopupCompletion = 2,
};

---@class QCompleter
---@field completionMode CompleterCompletionMode The completion mode.
local QCompleter = {}

---Creates a new Completer.
---@param params string[] The list of suggestions.
---@return QCompleter completer The new Completer.
function qt.QCompleter.create(params) end

---Returns current completion.
---@return string
function qt.QCompleter:currentCompletion() end
---@param callback function The function to be called when user choice is selected from popup.
function qt.QCompleter.onActivated(callback) end

---@class QClipboard A Lua wrapper for the Qt `QClipboard` class.
---@field text string The text content of the clipboard. Gets or sets the text content of the clipboard.
qt.QClipboard = {}

---Returns the global clipboard object.
---@return QClipboard globalClipboard The global clipboard object.
function qt.clipboard() end

---@enum TextElideMode
qt.TextElideMode = {
    ElideLeft = 0,
    ElideRight = 0,
    ElideMiddle = 0,
    ElideNone = 0,
}

qt.QDir = {
    ---@enum Filters
    Filters = {
        Dirs = 0,
        Files = 0,
        Drives = 0,
        NoSymLinks = 0,
        AllEntries = 0,
        TypeMask = 0,
        Readable = 0,
        Writable = 0,
        Executable = 0,
        PermissionMask = 0,
        Modified = 0,
        Hidden = 0,
        System = 0,
        AccessMask = 0,
        AllDirs = 0,
        CaseSensitive = 0,
        NoDot = 0,
        NoDotDot = 0,
        NoDotAndDotDot = 0,
        NoFilter = 0,
    },

    ---@enum SortFlags
    SortFlags = {
        Name = 0,
        Time = 0,
        Size = 0,
        Unsorted = 0,
        SortByMask = 0,
        DirsFirst = 0,
        Reversed = 0,
        IgnoreCase = 0,
        DirsLast = 0,
        LocaleAware = 0,
        Type = 0,
        NoSort = 0,
    }
}

qt.QDirIterator = {
    ---@enum IteratorFlag
    IteratorFlag = {
        NoIteratorFlags = 0,
        FollowSymlinks = 0,
        Subdirectories = 0,
    }
}

qt.QFileDevice = {
    ---@enum Permission
    Permission = {
        ReadOwner = 0,
        ReadUser = 0,
        ReadGroup = 0,
        ReadOther = 0,
        WriteOwner = 0,
        WriteUser = 0,
        WriteGroup = 0,
        WriteOther = 0,
        ExeOwner = 0,
        ExeUser = 0,
        ExeGroup = 0,
        ExeOther = 0,
    }
}

qt.QStandardPaths = {
    ---@enum StandardLocation
    StandardLocation = {
        DesktopLocation = 0,
        DocumentsLocation = 0,
        FontsLocation = 0,
        ApplicationsLocation = 0,
        MusicLocation = 0,
        MoviesLocation = 0,
        PicturesLocation = 0,
        TempLocation = 0,
        HomeLocation = 0,
        AppLocalDataLocation = 0,
        CacheLocation = 0,
        GenericDataLocation = 0,
        RuntimeLocation = 0,
        ConfigLocation = 0,
        DownloadLocation = 0,
        GenericCacheLocation = 0,
        GenericConfigLocation = 0,
        AppDataLocation = 0,
        AppConfigLocation = 0,
        PublicShareLocation = 0,
        TemplatesLocation = 0,
    }
}

return qt

File: src/plugins/lua/bindings/qt.cpp
// Copyright (C) 2024 The Qt Company Ltd.
// SPDX-License-Identifier: LicenseRef-Qt-Commercial OR GPL-3.0-only WITH Qt-GPL-exception-1.0

#include "../luaengine.h"

#include "utils.h"

#include <QApplication>
#include <QClipboard>
#include <QCompleter>
#include <QDir>
#include <QFileDevice>
#include <QStandardPaths>

namespace Lua::Internal {

void setupQtModule()
{
    registerProvider("Qt", [](sol::state_view lua) {
        sol::table qt(lua, sol::create);
        const ScriptPluginSpec *pluginSpec = lua.get<ScriptPluginSpec *>("PluginSpec");

        qt.new_usertype<QCompleter>(
            "QCompleter",
            "create",
            [](const QStringList &list) -> std::unique_ptr<QCompleter> {
                return std::make_unique<QCompleter>(list);
            },
            "currentCompletion",
            &QCompleter::currentCompletion,
            "completionMode",
            sol::property(
                &QCompleter::completionMode,
                [](QCompleter *c, QCompleter::CompletionMode mode) { c->setCompletionMode(mode); }),
            "onActivated",
            sol::property([guard = pluginSpec](QCompleter &obj, sol::function callback) {
                QObject::connect(
                    &obj,
                    QOverload<const QString &>::of(&QCompleter::activated),
                    guard->connectionGuard.get(),
                    [callback](const QString &arg) { void_safe_call(callback, arg); });
            }));

        qt.new_usertype<QClipboard>(
            "QClipboard",
            sol::no_constructor,
            "text",
            sol::property(
                [](QClipboard &self) { return self.text(); },
                [](QClipboard &self, const QString &value) { self.setText(value); }));

        qt["clipboard"] = &QApplication::clipboard;

        mirrorEnum(qt, QMetaEnum::fromType<QCompleter::CompletionMode>(), "QCompleterCompletionMode");

        // clang-format off
        qt["TextElideMode"] = lua.create_table_with(
            "ElideLeft", Qt::ElideLeft,
            "ElideRight", Qt::ElideRight,
            "ElideMiddle", Qt::ElideMiddle,
            "ElideNone", Qt::ElideNone
        );

        qt["QDirIterator"] = lua.create_table_with(
            "IteratorFlag", lua.create_table_with(
                "NoIteratorFlags", QDirIterator::NoIteratorFlags,
                "FollowSymlinks", QDirIterator::FollowSymlinks,
                "Subdirectories", QDirIterator::Subdirectories
            )
        );

        qt["QDir"] = lua.create_table_with(
            // QDir::Filters
            "Filters", lua.create_table_with(
                "Dirs", QDir::Dirs,
                "Files", QDir::Files,
                "Drives", QDir::Drives,
                "NoSymLinks", QDir::NoSymLinks,
                "AllEntries", QDir::AllEntries,
                "TypeMask", QDir::TypeMask,
                "Readable", QDir::Readable,
                "Writable", QDir::Writable,
                "Executable", QDir::Executable,
                "PermissionMask", QDir::PermissionMask,
                "Modified", QDir::Modified,
                "Hidden", QDir::Hidden,
                "System", QDir::System,
                "AccessMask", QDir::AccessMask,
                "AllDirs", QDir::AllDirs,
                "CaseSensitive", QDir::CaseSensitive,
                "NoDot", QDir::NoDot,
                "NoDotDot", QDir::NoDotDot,
                "NoDotAndDotDot", QDir::NoDotAndDotDot,
                "NoFilter", QDir::NoFilter
            ),

            // QDir::SortFlag
            "SortFlags", lua.create_table_with(
                "Name", QDir::Name,
                "Time", QDir::Time,
                "Size", QDir::Size,
                "Unsorted", QDir::Unsorted,
                "SortByMask", QDir::SortByMask,
                "DirsFirst", QDir::DirsFirst,
                "Reversed", QDir::Reversed,
                "IgnoreCase", QDir::IgnoreCase,
                "DirsLast", QDir::DirsLast,
                "LocaleAware", QDir::LocaleAware,
                "Type", QDir::Type,
                "NoSort", QDir::NoSort
            )
        );

        qt["QFileDevice"] = lua.create_table_with(
            "Permission", lua.create_table_with(
                "ReadOwner", QFileDevice::ReadOwner,
                "ReadUser", QFileDevice::ReadUser,
                "ReadGroup", QFileDevice::ReadGroup,
                "ReadOther", QFileDevice::ReadOther,
                "WriteOwner", QFileDevice::WriteOwner,
                "WriteUser", QFileDevice::WriteUser,
                "WriteGroup", QFileDevice::WriteGroup,
                "WriteOther", QFileDevice::WriteOther,
                "ExeOwner", QFileDevice::ExeOwner,
                "ExeUser", QFileDevice::ExeUser,
                "ExeGroup", QFileDevice::ExeGroup,
                "ExeOther", QFileDevice::ExeOther
            )
        );

        qt["QStandardPaths"] = lua.create_table_with(
            "StandardLocation", lua.create_table_with(
                "DesktopLocation", QStandardPaths::DesktopLocation,
                "DocumentsLocation", QStandardPaths::DocumentsLocation,
                "FontsLocation", QStandardPaths::FontsLocation,
                "ApplicationsLocation", QStandardPaths::ApplicationsLocation,
                "MusicLocation", QStandardPaths::MusicLocation,
                "MoviesLocation", QStandardPaths::MoviesLocation,
                "PicturesLocation", QStandardPaths::PicturesLocation,
                "TempLocation", QStandardPaths::TempLocation,
                "HomeLocation", QStandardPaths::HomeLocation,
                "AppLocalDataLocation", QStandardPaths::AppLocalDataLocation,
                "CacheLocation", QStandardPaths::CacheLocation,
                "GenericDataLocation", QStandardPaths::GenericDataLocation,
                "RuntimeLocation", QStandardPaths::RuntimeLocation,
                "ConfigLocation", QStandardPaths::ConfigLocation,
                "DownloadLocation", QStandardPaths::DownloadLocation,
                "GenericCacheLocation", QStandardPaths::GenericCacheLocation,
                "GenericConfigLocation", QStandardPaths::GenericConfigLocation,
                "AppDataLocation", QStandardPaths::AppDataLocation,
                "AppConfigLocation", QStandardPaths::AppConfigLocation,
                "PublicShareLocation", QStandardPaths::PublicShareLocation,
                "TemplatesLocation", QStandardPaths::TemplatesLocation
        ));
        // clang-format on

        return qt;
    });
}

} // namespace Lua::Internal
